src/fpm.f90

```diff
@@ -203,6 +192,138 @@ subroutine build_model(model, settings, package, error)
+                        if (file_scope_flag.eq."") then
+                            sources(i)%flags=model%cmd_compile_flags//" "//profile%flags
+                        else
+                            sources(i)%flags=model%cmd_compile_flags//" "//file_scope_flag
+                        end if
+                    end if
+                case (FPM_UNIT_PROGRAM)
+                    file_scope_flag = get_file_scope_flags(sources(i), profile)
+                    if (file_scope_flag.eq."") then
+                        sources(i)%flags=model%cmd_compile_flags//" "//profile%flags
+                    else
+                        sources(i)%flags=model%cmd_compile_flags//" "//file_scope_flag
+                    end if
+                    sources(i)%link_time_flags=profile%link_time_flags
+                end select
+            end do
+        end associate
+    end do
+
+    contains
+
+    ! Look for an appropriate profile
+    ! If package has specified profile, return it
+    ! If it has just built-in profile, try to find specified one in parents, otherwise return it
+    ! If it has no profiles, try to find one in parents
+    function look_for_profile(package_id) result (chosen_profile)
+        integer, intent(in) :: package_id
+
+        integer :: idx
+        type(profile_config_t), allocatable :: built_in, chosen_profile
+        type(profile_config_t) :: current
+        logical :: profile_found
+
+        idx = package_id
+        associate(pkgs => model%packages)
+            do while (.true.)
+                profile_found = .false.
+                if (allocated(pkgs(idx)%profiles)) then
+                    call find_profile(pkgs(idx)%profiles, profile, model%fortran_compiler, &
+                        & get_os_type(), profile_found, current)
+                    if (profile_found) then
+                        if (current%is_built_in) then
+                            if (.not. allocated(built_in)) then
+                                built_in = current
+                                chosen_profile = current
+                            end if
+                            if (allocated(pkgs(idx)%parent)) then
+                                idx = pkgs(idx)%parent(1)
+                            else
+                                exit
+                            end if
+                        else
+                            chosen_profile = current
+                            return
+                        end if
+                    end if
+                else
+                    if (allocated(pkgs(idx)%parent)) then
+                        idx = pkgs(idx)%parent(1)
+                    else
+                        call fpm_stop(1,'*look_for_profile*:Error: Orphan package does not have any profiles.')
+                    end if
+                end if
+            end do
+        end associate
+        if (.not. allocated(chosen_profile)) call fpm_stop(1,'*look_for_profile*:Error: No profile found.')
```

Comment:
Is there a situation where this error would occur, or should this never happen? Same question for line 296

Reply:
The error on 301 should probably happen only if the main package does not specify or has built-in the correct profile.
296 should never happen as there should always be a path from any dependency to main package.